rust/src/transformations/make_stable_expr/expr_clip/mod.rs

```diff
@@ -0,0 +1,158 @@
+use polars::prelude::*;
+use polars_plan::dsl::{Expr, FunctionExpr};
+
+use crate::core::{ExprFunction, Function, MetricSpace, StabilityMap, Transformation};
+use crate::domains::{AtomDomain, Bounds, ExprDomain, NumericDataType, SeriesDomain};
+use crate::error::*;
+use crate::traits::Number;
+
+use super::{DatasetOuterMetric, StableExpr};
+
+/// Make a Transformation that returns a `clip(bounds)` expression for a LazyFrame.
+///
+/// # Arguments
+/// * `input_domain` - Expr domain
+/// * `input_metric` - The metric under which neighboring LazyFrames are compared
+/// * `expr` - The clipping expression
+pub fn make_expr_clip<MS, MI>(
+    input_domain: ExprDomain,
+    input_metric: MS,
+    expr: Expr,
+) -> Fallible<Transformation<ExprDomain, ExprDomain, MS, MI>>
+where
+    MS: 'static + DatasetOuterMetric,
+    MI: 'static + DatasetOuterMetric,
+    MI::Distance: Clone,
+    (ExprDomain, MS): MetricSpace,
+    (ExprDomain, MI): MetricSpace,
+    Expr: StableExpr<MS, MI>,
+{
+    let Expr::Function {
+        input, function, ..
+    } = expr
+    else {
+        return fallible!(MakeTransformation, "expected function expression");
+    };
+
+    let FunctionExpr::Clip { has_min, has_max } = function else {
+        return fallible!(MakeTransformation, "expected clip function");
+    };
```

Comment:
Errors from these matches should be completely internal and never seen by a library user. I could unpack these arguments out in the trait impl for StableExpr, and pass them in, but didn't do so because I wanted to keep the trait impl dispatching as small as possible. 

I'm a little worried about duplicate code in the trait impls for StableExpr, as well, so the less code there the better. Duplicate code/macros may arise because there ultimately needs to be a trait impl for each combination of input/output metric/measure.